airbyte-config/models/src/main/java/io/airbyte/config/helpers/LogClientSingleton.java

```diff
@@ -19,95 +18,96 @@
 import org.slf4j.LoggerFactory;
 import org.slf4j.MDC;
 
-// todo (cgardens) - make this an actual singleton so we can write tests and mock the components.
 /**
  * Airbyte's logging layer entrypoint. Handles logs written to local disk as well as logs written to
  * cloud storages.
- *
+ * <p>
  * Although the configuration is passed in as {@link Configs}, it is transformed to
  * {@link LogConfigs} within this class. Beyond this class, all configuration consumption is via the
  * {@link LogConfigs} interface via the {@link CloudLogs} interface.
  */
 public class LogClientSingleton {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(LogClientSingleton.class);
+  private static LogClientSingleton instance;
 
   @VisibleForTesting
-  static final int LOG_TAIL_SIZE = 1000000;
+  final static int LOG_TAIL_SIZE = 1000000;
   @VisibleForTesting
-  static CloudLogs logClient;
+  CloudLogs logClient;
```

Comment:
What motivation is there for making it static?  Right now it's being initialized in a weird roundabout way from any method that might go to use it, and there's no guarantees as to which method got its initialization through first.  That makes code harder to reason about. I think this should not only not be static, but shouldn't even be a member variable.  Any method that needs it should create it in local scope only.  That way there are solid guarantees about how it's constructed/fetched/initialized, and the code paths are simpler. 

I started digging into this to figure out why it would ever be static to begin with, and the code trail takes me to other classes that seem to be using static patterns unnecessarily, and with no clear entry point for how stuff is reliably initialized, making the paths even more opaque.  I think there's another refactor possible to improve the API on those and make their entry points and configuration more clear, more guaranteed, and more testable.

Reply:
Client initialisation can take some time with these Storage clients since they are meant to be used concurrently and often have multiple connections and buffers under the hood. When I was writing this, I was wary about creating too many in memory. Taking a second look at the code, this should be fine here since each individual implementation has it's own static client.

In general, I set things to static as I do not want to need to reason about state within an object. It is something I look out for when reading Java code. Static is something that is often set once and forget, while member variables are something I have to keep in mind as they might/might not be mutated. This cascades down into functions too.

My bad for naming this as singleton since it's not a strict singleton implementation - what I was going for here is more of a complete static class. There isn't a need for any member variables since the object isn't manipulating any internal state (any methods here should not modify the logClient's state or change behavior depending on it's state).

One thing we can probably do to clear this specific class up is eagerly initialising the `logClient` instead of lazy initialising it. You are right that lazy initialising doesn't provide much benefit here. We are going to use the client regardless and we don't save much by deferring the construction to later. Pretty simple path here of 1) removing the lazy calls 2) call the `createCloudClientIfNull` always and set the result to the static `logClient` variable 3) if we want to we can even get rid of the singleton `getInstance` pattern, since this the logic here is stateless in general, though I think keeping it is fine.

The proper singleton pattern might apply more to the individual Cloud Storage clients. 

It's probably easier to talk through this since I'm realising this involves the log client implementations and it more efficient to discuss them altogether. I scheduled a meeting with you, please feel free to move it.

Reply:
Discussed offline.  We can leave this for now for this PR.  I expect a later refactor to address a bunch of class structure concerns around logging, and can wrap in the CloudLogs references in that, as there are a number of classes collaborating in logging config and access and I'll tackle them as a set.